sshutil/templates.go

```diff
@@ -0,0 +1,195 @@
+		ExtensionsKey: DefaultExtensions(ct),
+	}
+}
+
+// DefaultExtensions returns the default extensions set in an SSH certificate.
+func DefaultExtensions(ct CertType) map[string]interface{} {
+	switch ct {
+	case UserCert:
+		return map[string]interface{}{
+			"permit-X11-forwarding":   "",
+			"permit-agent-forwarding": "",
+			"permit-port-forwarding":  "",
+			"permit-pty":              "",
+			"permit-user-rc":          "",
+		}
+	default:
+		return nil
+	}
+}
+
+// NewTemplateData creates a new map for templates data.
+func NewTemplateData() TemplateData {
+	return TemplateData{}
+}
+
+// AddExtension adds one extension to the templates data.
+func (t TemplateData) AddExtension(key, value string) {
+	if m, ok := t[ExtensionsKey].(map[string]interface{}); ok {
+		m[key] = value
+	} else {
+		t[ExtensionsKey] = map[string]interface{}{
+			key: value,
+		}
+	}
+}
+
+// AddCriticalOption adds one critical option to the templates data.
+func (t TemplateData) AddCriticalOption(key, value string) {
+	if m, ok := t[CriticalOptionsKey].(map[string]interface{}); ok {
+		m[key] = value
+	} else {
+		t[CriticalOptionsKey] = map[string]interface{}{
+			key: value,
+		}
+	}
+}
+
+// Set sets a key-value pair in the template data.
+func (t TemplateData) Set(key string, v interface{}) {
+	t[key] = v
+}
+
+// SetInsecure sets a key-value pair in the insecure template data.
+func (t TemplateData) SetInsecure(key string, v interface{}) {
+	if m, ok := t[InsecureKey].(TemplateData); ok {
+		m[key] = v
+	} else {
+		t[InsecureKey] = TemplateData{key: v}
+	}
+}
+
+// SetType sets the certificate type in the template data.
+func (t TemplateData) SetType(typ CertType) {
+	t.Set(TypeKey, typ.String())
+}
+
+// SetKeyID sets the certificate key id in the template data.
+func (t TemplateData) SetKeyID(id string) {
+	t.Set(KeyIDKey, id)
+}
+
+// SetPrincipals sets the certificate principals in the template data.
+func (t TemplateData) SetPrincipals(p []string) {
+	t.Set(PrincipalsKey, p)
+}
+
+// SetExtensions sets the certificate extensions in the template data.
+func (t TemplateData) SetExtensions(e map[string]interface{}) {
+	t.Set(ExtensionsKey, e)
+}
+
+// SetCriticalOptions sets the certificate critical options in the template
+// data.
+func (t TemplateData) SetCriticalOptions(o map[string]interface{}) {
+	t.Set(CriticalOptionsKey, o)
+}
+
+// SetToken sets the given token in the template data.
+func (t TemplateData) SetToken(v interface{}) {
+	t.Set(TokenKey, v)
+}
+
+// SetUserData sets the given user provided object in the insecure template
+// data.
+func (t TemplateData) SetUserData(v interface{}) {
+	t.SetInsecure(UserKey, v)
+}
+
+// SetCertificateRequest sets the simulated ssh certificate request the insecure
+// template data.
+func (t TemplateData) SetCertificateRequest(cr CertificateRequest) {
+	t.SetInsecure(CertificateRequestKey, cr)
+}
+
+// DefaultCertificate is the default template for an SSH certificate.
+const DefaultCertificate = `{
```

Comment:
shouldn't this name be DefaultTemplate? otherwise the name will be sshutil.DefaultCertificate, which sort of implies  something different.

Reply:
Yes, in fact in x509util use use things like `DefaultLeafTemplate`